optaplanner-examples/src/main/java/org/optaplanner/examples/flightcrewscheduling/optional/score/FlightCrewSchedulingConstraintProvider.java

```diff
@@ -0,0 +1,63 @@
+package org.optaplanner.examples.flightcrewscheduling.optional.score;
+
+import java.util.Objects;
+
+import org.optaplanner.core.api.score.buildin.hardsoftlong.HardSoftLongScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.examples.flightcrewscheduling.domain.Employee;
+import org.optaplanner.examples.flightcrewscheduling.domain.FlightAssignment;
+
+public class FlightCrewSchedulingConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[]{
+                flightConflict(constraintFactory),
+                requiredSkill(constraintFactory),
+                employeeUnavailability(constraintFactory),
+                transferBetweenTwoFlights(constraintFactory),
+                firstAssignmentDepartingFromHome(constraintFactory),
+                lastAssignmentArrivingAtHome(constraintFactory)
+        };
+    }
+
+    private Constraint flightConflict(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(FlightAssignment.class)
+                .filter((first, second) -> Objects.equals(first.getEmployee(), second.getEmployee()))
+                .filter((first, second) -> second.getFlight().overlaps(first.getFlight()))
+                .filter((first, second) -> second.getId() > first.getId())
+                .penalize("Flight conflict", HardSoftLongScore.ofHard(10));
+    }
+
+    private Constraint requiredSkill(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(FlightAssignment.class)
+                .filter(a -> !a.getEmployee().hasSkill(a.getRequiredSkill()))
+                .penalize("Required skill", HardSoftLongScore.ofHard(100));
+    }
+
+    private Constraint employeeUnavailability(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(FlightAssignment.class)
+                .filter(a -> !a.getEmployee().isAvailable(a.getFlight().getDepartureUTCDate()))
+                .penalize("Employee unavailable", HardSoftLongScore.ofHard(10));
+    }
+
+    private Constraint transferBetweenTwoFlights(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(Employee.class)
+                .filter(e -> e.countInvalidConnections() > 0)
```

Comment:
Can we use the variable name "employee" in such cases? (Same for a above constraints)

lamdba's don't have an explicit type written down. (It says "e", not "Employee e"), so for readability (most users start from one of these examples), let's use a full variable name.

In general, in optaplanner-core only the following short var names are used: int i, Object o, Exception e.